[CGData][MachineOutliner] Global Outlining

kyulee-com opened the pull request:
This commit introduces support for outlining functions across modules using codegen data generated from previous codegen. The codegen data currently manages the outlined hash tree, which records outlining instances that occurred locally in the past.
    
The machine outliner now operates in one of three modes:

1.  CGDataMode::None: This is the default outliner mode that uses the suffix tree to identify (local) outlining candidates within a module. This mode is also used by (full)LTO to maintain optimal behavior with the combined module.
2. CGDataMode::Write (`-codegen-data-generate`): This mode is identical to the default mode, but it also publishes the stable hash sequences of instructions in the outlined functions into a local outlined hash tree. It then encodes this into the `__llvm_outline` section, which will be dead-stripped at link time.
3. CGDataMode::Read (`-codegen-data-use-path={.cgdata}`): This mode reads a codegen data file (.cgdata) and initializes a global outlined hash tree. This tree is used to generate global outlining candidates. Note that the codegen data file has been post-processed with the raw `__llvm_outline` sections from all native objects using the `llvm-cgdata` tool (or a linker, `LLD`, or a new ThinLTO pipeline later).

This depends on https://github.com/llvm/llvm-project/pull/105398. After this PR, LLD (https://github.com/llvm/llvm-project/pull/90166) and Clang (https://github.com/llvm/llvm-project/pull/90304) will follow for each client side support.
This is a patch for https://discourse.llvm.org/t/rfc-enhanced-machine-outliner-part-2-thinlto-nolto/78753.

kyulee-com committed:
[CGData][MachineOutliner] Global Outlining2

This commit introduces support for outlining functions across modules using codegen data generated from previous codegen. The codegen data currently manages the outlined hash tree, which records outlining instances that occurred locally in the past.

The machine outliner now operates in one of three modes:
1. CGDataMode::None: This is the default outliner mode that uses the suffix tree to identify (local) outlining candidates within a module. This mode is also used by (full)LTO to maintain optimal behavior with the combined module.
2. CGDataMode::Write (`codegen-data-generate`): This mode is identical to the default mode, but it also publishes the stable hash sequences of instructions in the outlined functions into a local outlined hash tree. It then encodes this into the `__llvm_outline` section, which will be dead-stripped at link time.
3. CGDataMode::Read (`codegen-data-use-path={.cgdata}`): This mode reads a codegen data file (.cgdata) and initializes a global outlined hash tree. This tree is used to generate global outlining candidates. Note that the codegen data file has been post-processed with the raw `__llvm_outline` sections from all native objects using the `llvm-cgdata` tool (or a linker, `LLD`, or a new ThinLTO pipeline later).

kyulee-com commented on llvm/lib/CodeGen/MachineOutliner.cpp:
```diff
@@ -128,6 +135,19 @@ static cl::opt<bool> OutlinerLeafDescendants(
              "tree as candidates for outlining (if false, only leaf children "
              "are considered)"));
 
+static cl::opt<bool>
```
These flags are currently added for testing purposes and are not strictly necessary. However, if everyone agrees, I am happy to remove them.

aemerson commented on llvm/lib/CodeGen/MachineOutliner.cpp:
```diff
@@ -585,6 +644,111 @@ void MachineOutliner::emitOutlinedFunctionRemark(OutlinedFunction &OF) {
+  assert(cgdata::hasOutlinedHashTree());
+  const auto *RootNode = cgdata::getOutlinedHashTree()->getRoot();
+  for (size_t I = 0; I < Size; ++I) {
+    // skip the invalid mapping that represents a large negative value.
```
Capitalize the first letter in comments (also the others in this function)

but also, what's this large negative value meaning?

kyulee-com replied:
Thanks for the catch! The large negative value (starting from `-3`) in the `InstructionMapper` indicates that an instruction mapping is not valid or legal -- https://github.com/kyulee-com/llvm-project/blob/master/llvm/lib/CodeGen/MachineOutliner.cpp#L216-L217, which I implicitly relied on. Instead, to clarify the code, I now check a valid mapping with `Mapper.LegalInstrNumber`.

aemerson commented on llvm/lib/CodeGen/MachineOutliner.cpp:
```diff
@@ -695,6 +859,39 @@ void MachineOutliner::findCandidates(
+  if (OutlinerMode == CGDataMode::Write) {
+    StableHashAttempts++;
+    if (!OutlinedHashSequence.empty())
+      LocalHashTree->insert({OutlinedHashSequence, CandSize});
```
Why do we need to store the number of candidates for this already-outlined function?

kyulee-com replied:
We track the frequency of each outlined hash sequence in the LocalHashTree (in write mode), which helps us understand how often each sequence appears in a specific module. This frequency is represented as `Terminals` -- https://github.com/llvm/llvm-project/blob/main/llvm/lib/CGData/OutlinedHashTree.cpp#L87.
When combining data from all modules, this terminal count shows the total occurrences of each sequence across all modules.

In this PR, when we create a `GlobalOutlinedFunction` using CGData from previous runs (in read mode), we initialize a `GlobalOccurrenceCount` from the `Terminals` in the global hash tree, which has been merged. This count helps us decide which candidates are most important to outline based on their frequency across all modules.

aemerson replied:
> This count helps us decide which candidates are most important to outline based on their frequency across all modules.

This is I guess the core of my question. As long as you have a size saving, surely you would outline *all* profitable candidate functions rather so ordering them seems redundant? I'm probably missing something.

kyulee-com replied:
We create an outlining candidate each time a sequence of instructions matches in the global hash tree. These candidates may overlap. As shown in the `outline` function (https://github.com/llvm/llvm-project/blob/main/llvm/lib/CodeGen/MachineOutliner.cpp#L837-L868), we sort these candidates and process them sequentially, discarding any that overlap with previously processed ones. Initially, our global outlining implementation did not utilize this count information. However, we discovered that prioritizing candidates based on their frequency significantly improves size efficiency in real-world workloads.

Although it's theoretically possible to globally order all candidates using the hash tree, we can't access to other modules to determine the best outlining sequence, as modules are fundamentally independent. In this PR, we populate candidates that appear in the current module, match them in the global hash tree, and then order these local instances based on a global heuristic. While this may not be optimal, it has proven effective in our tests.

For more details, you can refer to a unit test related to candidate prioritization in this PR: `llvm/test/CodeGen/AArch64/cgdata-read-priority.ll`

kyulee-com committed:
Address comments from aemerson

- Capitalize the comments
- Use Mapper.LegalInstrNumber
- Remove dead code (Sequence)

kyulee-com commented:
Could someone please take another look at this code? It is a crucial component for integrating the global outlining capability into the machine outliner. I've made efforts to ensure that the integration disrupts the existing logic as little as possible. Thanks!

ellishg reviewed (commented):
Overall this looks good to me. Thanks for adding thorough tests!

kyulee-com committed:
Address comments from ellishg

kyulee-com commented on llvm/lib/CodeGen/MachineOutliner.cpp:
```diff
@@ -585,6 +644,107 @@ void MachineOutliner::emitOutlinedFunctionRemark(OutlinedFunction &OF) {
+      continue;
+
+    size_t J = I + 1;
+    for (; J < Size; ++J) {
```
Nope. It's slipped from refactoring. I combined them.

kyulee-com commented on llvm/lib/CodeGen/MachineOutliner.cpp:
```diff
@@ -585,6 +644,107 @@ void MachineOutliner::emitOutlinedFunctionRemark(OutlinedFunction &OF) {
+        break;
+      LastNode = followHashNode(StableHashJ, LastNode);
+      if (!LastNode)
+        break;
```
Yeap. I restructured the loops.

kyulee-com commented on llvm/lib/CodeGen/MachineOutliner.cpp:
```diff
@@ -585,6 +644,107 @@ void MachineOutliner::emitOutlinedFunctionRemark(OutlinedFunction &OF) {
+      // Even with a match ending with a terminal, we continue finding
+      // matches to populate all candidates.
+      auto Count = LastNode->Terminals;
+      if (Count)
```
Combined them into a single line.

kyulee-com commented on llvm/lib/CodeGen/MachineOutliner.cpp:
```diff
@@ -585,6 +644,107 @@ void MachineOutliner::emitOutlinedFunctionRemark(OutlinedFunction &OF) {
+// Find all matches in the global outlined hash tree.
+// It's quadratic complexity in theory, but it's nearly linear in practice
+// since the length of outlined sequences are small within a block.
+static std::vector<MatchedEntry> getMatchedEntries(InstructionMapper &Mapper) {
```
Switched to `SmallVector<>`.

kyulee-com commented on llvm/lib/CodeGen/MachineOutliner.cpp:
```diff
@@ -585,6 +644,107 @@ void MachineOutliner::emitOutlinedFunctionRemark(OutlinedFunction &OF) {
+                                       MinRepeats);
+    if (!OF.has_value() || OF.value()->Candidates.empty())
+      continue;
+    // We create a global candidate each match.
```
Fixed.

kyulee-com commented:
I think I've addressed all the comments so far. Could someone please take another look? Thanks!

ellishg reviewed (approved):
LGTM thanks for fixing my nits

kyulee-com commented:
> Looks like this is breaking llvm/test/CodeGen/Generic/llc-start-stop.ll on macOS: http://45.33.8.238/macm1/91798/step_10.txt
> 
> Please take a look and revert for now if it takes a while to fix.

Thanks for the head-up! I made a forward fix with https://github.com/llvm/llvm-project/pull/108037. Hope this resolves this case, too.

nico commented:
The forward fix broke check-llvm on linux too: http://45.33.8.238/linux/147290/step_11.txt

nico commented:
I don't know, it hasn't landed yet. Go ahead and commit I suppose – trunk is already broken, it won't get worse. But if things are still broken then, it's probably time to regroup, revert this and the then-two follow-ups and find a local repro to confirm fixes on.

kyulee-com commented:
> I don't know, it hasn't landed yet. Go ahead and commit I suppose – trunk is already broken, it won't get worse. But if things are still broken then, it's probably time to regroup, revert this and the then-two follow-ups and find a local repro to confirm fixes on.

Sure, I will land it first. If things are not fixed, I will revert the entire change.
By the way, could you approve https://github.com/llvm/llvm-project/pull/108047? Somehow, I can't directly commit it to the main branch, although I have done so before.

nico commented:
check-llvm is still broken: http://45.33.8.238/linux/147291/step_11.txt
